craft_parts/sources/base.py

```diff
@@ -0,0 +1,217 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright 2015-2021 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Base classes for source type handling."""
+
+import abc
+import os
+import shutil
+from pathlib import Path
+from typing import List, Optional, Union
+
+import requests
+
+from craft_parts import utils
+from craft_parts.dirs import ProjectDirs
+from craft_parts.utils import url_utils
+
+from . import errors
+from .cache import FileCache
+from .checksum import verify_checksum
+
+
+class SourceHandler(abc.ABC):
+    """The base class for source type handlers.
+
+    Methods :meth:`check_if_outdated` and :meth:`update_source` can be
+    overridden by subclasses to implement verification and update of
+    source files.
+    """
+
+    # pylint: disable=too-many-arguments
+
+    def __init__(
+        self,
+        source: Union[str, Path],
+        source_dir: Union[str, Path],
+        *,
+        application_name: str = None,
```

Comment:
```suggestion
        application_name: Optional[str] = None,
```

and so on.

Reply:
It increases character density in the parameter list but it's more consistent with the optional parameter definition.